data/transactions/logic/eval.go

```diff
@@ -2119,6 +2159,49 @@ func opSwitch(cx *EvalContext) error {
 	return nil
 }
 
+func opMatch(cx *EvalContext) error {
+	n := int(cx.program[cx.pc+1])
+	// stack contains the n sized match list and the single match value
+	if n+1 > len(cx.stack) {
+		return fmt.Errorf("match expects %d stack args while stack only contains %d", n+1, len(cx.stack))
+	}
+
+	top := len(cx.stack) - 1
+	matchVal := cx.stack[top]
+	cx.stack = cx.stack[:top]
+
+	top = len(cx.stack) - n
+	matchList := cx.stack[top:]
+	cx.stack = cx.stack[:top]
+
+	// first we must verify that the matchList has the expected length and all types match the type of matchVal
+	if len(matchList) != n {
+		return fmt.Errorf("list of match constants has length %d when length %d was expected", len(matchList), n)
+	}
+
+	matchedIdx := n
+	for i, stackArg := range matchList {
+		if stackArg.argType() != matchVal.argType() {
```

Comment:
I'm disinclined to perform this check. I don't want to iterate through all the values at all once we have a match.  Why not jump as soon as one is equal?

(And, if we want that optimization, we should not even consider it an error for the types to mismatch on the values we _do_ check, since that's a weird spec - the types would have to match until we find a value match, but the rest could be the wrong type.)

Reply:
That a good point, I'll update this code to have more straightforward semantics.

Reply:
Is still don't like this initial type check.  We would have to spec that all the match values _before_ the found value must match types, but afterward, we don't care.

Reply:
Oh weird, I thought I changed this. My bad.

Reply:
Ok, I think that's correct, but given that your equality matching checks types, I don't think it accomplishes anything.